src/SlmMail/Service/SparkPostService.php

```diff
@@ -0,0 +1,241 @@
+     *
+     * @param string $apiKey
+     */
+    public function __construct($apiKey)
+    {
+        $this->apiKey = (string)$apiKey;
+    }
+
+
+    public function send(Message $message)
+    {
+
+        if ($message instanceof SparkPostMessage) {
+            $options = $message->getOptions();
+        }
+
+        $spec['api_key'] = $this->apiKey;
+
+        // Prepare message
+        $from = $this->prepareFromAddress($message);
+        $recipients = $this->prepareRecipients($message);
+        $headers = $this->prepareHeaders($message);
+        $body = $this->prepareBody($message);
+
+        $post = [
+            'options' => $options,
+            'content' => [
+                'from' => $from,
+                'subject' => $message->getSubject(),
+                'html' => $body,
+            ],
+        ];
+
+        $post = array_merge($post, $recipients);
+        if ((count($recipients) == 0) && (!empty($headers) || !empty($body))) {
+            throw new Exception\RuntimeException(
+                sprintf(
+                    '%s transport expects at least one recipient if the message has at least one header or body',
+                    __CLASS__
+                )
+            );
+        }
+
+        $response = $this->prepareHttpClient('/transmissions', $post)
+            ->send()
+        ;
+
+        return $this->parseResponse($response);
+    }
+
+    /**
+     * Retrieve email address for envelope FROM
+     *
+     * @param  Message $message
+     *
+     * @throws Exception\RuntimeException
+     * @return string
+     */
+    protected function prepareFromAddress(Message $message)
+    {
+        #if ($this->getEnvelope() && $this->getEnvelope()->getFrom()) {
+        #    return $this->getEnvelope()->getFrom();
+        #}
+
+        $sender = $message->getSender();
+        if ($sender instanceof Address\AddressInterface) {
+            return $sender->getEmail();
+        }
+
+        $from = $message->getFrom();
+        if (!count($from)) {
+            // Per RFC 2822 3.6
+            throw new Exception\RuntimeException(
+                sprintf(
+                    '%s transport expects either a Sender or at least one From address in the Message; none provided',
+                    __CLASS__
+                )
+            );
+        }
+
+        $from->rewind();
+        $sender = $from->current();
+
+        return $sender->getEmail();
+    }
+
+    /**
+     * Prepare array of email address recipients
+     *
+     * @param  Message $message
+     *
+     * @return array
+     */
+    protected function prepareRecipients(Message $message)
+    {
+        #if ($this->getEnvelope() && $this->getEnvelope()->getTo()) {
+        #    return (array) $this->getEnvelope()->getTo();
+        #}
+
+        $recipients = [];
+        $recipients['recipients'] = $this->prepareAddresses($message->getTo());
+        !($cc = $this->prepareAddresses($message->getCc())) || $recipients['cc'] = $cc;
```

Comment:
This is very weird to read. I suggest expanding it, to make it easier for other developers to read it, and understand what happens.